src/Types/TypedTxOut.purs

```diff
@@ -0,0 +1,289 @@
+  => FromData b
+  => ToData b
+  => TypedTxOutRef a b
+  -> TransactionInput
+typedTxOutRefInput (TypedTxOutRef { txOutRef }) = txOutRef
+
+-- A `TransactionOutput` tagged by a phantom type: and the connection type of
+-- the output. DO NOT import as extra constraints are required so only import
+-- the smart constructor `mkTypedTxOut`
+newtype TypedTxOut (a :: Type) (b :: Type) = TypedTxOut
+  { txOut :: TransactionOutput, data :: b }
+
+-- `DatumType a b` not needed but this replicates Plutus and provides extra
+-- type safety.
+derive newtype instance (DatumType a b, Eq b) => Eq (TypedTxOut a b)
+
+-- | Extract the `Address` of a `TypedTxOut`
+typedTxOutAddress
+  :: forall (a :: Type) (b :: Type)
+   . DatumType a b
+  => FromData b
+  => ToData b
+  => TypedTxOut a b
+  -> Address
+typedTxOutAddress (TypedTxOut { txOut }) = (unwrap txOut).address
+
+-- | Extract the `DatumHash` of a `TypedTxOut`
+typedTxOutDatumHash
+  :: forall (a :: Type) (b :: Type)
+   . DatumType a b
+  => FromData b
+  => ToData b
+  => TypedTxOut a b
+  -> Maybe DatumHash
+typedTxOutDatumHash (TypedTxOut { txOut }) = (unwrap txOut).data_hash
+
+-- | Extract the `Value` of a `TypedTxOut`
+typedTxOutValue
+  :: forall (a :: Type) (b :: Type)
+   . DatumType a b
+  => FromData b
+  => ToData b
+  => TypedTxOut a b
+  -> Value
+typedTxOutValue (TypedTxOut { txOut }) = (unwrap txOut).amount
+
+-- | Extract the `TxOut` ~ `TransactionOutput` of a `TypedTxOut`
+typedTxOutTxOut
+  :: forall (a :: Type) (b :: Type)
+   . DatumType a b
+  => FromData b
+  => ToData b
+  => TypedTxOut a b
+  -> TransactionOutput
+typedTxOutTxOut (TypedTxOut { txOut }) = txOut
+
+-- Purescript's lack of dependent types requires this smart constructor as we
+-- cannot constrain the `TypedTxOut` datatype.
+-- | Smart constructor to create a `TypedTxOut` from a network ID,
+-- | a correctly-typed data, script, an address, and a value. A smart
+-- | constructor is required because extra constraints are needed.
+-- | `TransactionOutput` is tagged by a phantom type.
+mkTypedTxOut
+  :: forall (a :: Type) (b :: Type)
+   . DatumType a b
+  => FromData b
+  => ToData b
+  => NetworkId
+  -> TypedValidator a
+  -> b
+  -> Value
+  -> Maybe (TypedTxOut a b)
+mkTypedTxOut networkId typedVal dt amount = do
+  dHash <- datumHash $ Datum $ toData dt
+  let address = typedValidatorAddress networkId typedVal
+  pure $ mkTypedTxOut' (wrap { address, amount, data_hash: pure dHash }) dt
+  where
+  mkTypedTxOut'
+    :: TransactionOutput
+    -> b -- Data
+    -> TypedTxOut a b
+  mkTypedTxOut' txOut dat = TypedTxOut { txOut, data: dat }
+
+data WrongOutTypeError = ExpectedScriptGotPubkey -- | ExpectedPubkeyGotScript
+
+derive instance Generic WrongOutTypeError _
+derive instance Eq WrongOutTypeError
+derive instance Ord WrongOutTypeError
+
+instance Show WrongOutTypeError where
+  show = genericShow
+
+-- | An error we can get while trying to type an existing transaction part.
+data ConnectionError
```

Comment:
Is this name taken from Plutus? If so, I don't think we should follow it. Especially in the context of CTL, I think the name would be take to mean a failure in some networked code

Reply:
It is, I will rename.